src/rgw/rgw_sal_motr.cc

```diff
@@ -1378,9 +1462,31 @@ void MotrObject::gen_rand_obj_instance_name()
 {
   enum {OBJ_INSTANCE_LEN = 32};
   char buf[OBJ_INSTANCE_LEN + 1];
-
+  std::string version_id;
   gen_rand_alphanumeric_no_underscore(store->ctx(), buf, OBJ_INSTANCE_LEN);
-  key.set_instance(buf);
+  //TODO: Handle null version object case in PutObj operation.
+  // As the version ID timestamp is encoded in Base62, the maximum value
+  // for 8-characters is 62^8 - 1. This is the maximum time interval in ms.
+  constexpr uint64_t max_ts_count = 218340105584895;
+  constexpr size_t ts_len = 8, rand_id_len = 24;
+  using UnsignedMillis = std::chrono::duration<uint64_t, std::milli>;
+  const auto ms_since_epoch = std::chrono::time_point_cast<UnsignedMillis>(
+                              std::chrono::system_clock::now()).time_since_epoch().count();
+  uint64_t cur_time = max_ts_count - ms_since_epoch;
+  auto version_ts = base62_encode(cur_time, ts_len);
+  // encode random uuid string.
+  auto uuid_str = base64_encode(reinterpret_cast<const unsigned char*>(buf), rand_id_len);
+  // find and replace all "=" with "";
+  std::string search_str = "=", replace_str = "";
+  for (size_t pos = uuid_str.find(search_str, 0);
+    pos != std::string::npos && search_str.length();
+    pos = uuid_str.find(search_str, pos + replace_str.length()))
+      uuid_str.replace(pos, search_str.length(), replace_str);
+
+  version_id = version_ts + uuid_str;
```

Comment:
Add reasoning here for future reference on why we are introducing our own way of generating version-id and not using how RGW generates it as random string, and for some reason sorting for those version-ids not working in expected lexicographic order

Reply:
Just to be clear, the lexicographic ordering is a by-product of the Motr backend, there is no user-facing reason to have any ordering. In fact, this is a downside to this implementation since the version IDs reveal implementation details and properties of the object. Completely random IDs (as done by Min.io, or presumably by standard Ceph (?), or AWS (as far as we know)) would be preferable.

Reply:
> As per AWS and CEPH documentation, the version-ids are listed in lexicographically sorted way.

Can you be more specific? I am not seeing anything in the links you provided that would indicate AWS or Ceph generate lexicographical sorted version IDs. Are we talking about different things?

My comment was that the version IDs returned by Amazon AWS have no particular ordering properties. The object version listing returns version sorted by time, not ID. The sorted version ID is specific to CORTX.

For example, here's the output of storing 10 objects into a version-enabled Amazon AWS bucket (listed using `mc`):

```text
❯ mc ls --versions s3/kpine-feebf0c1
[2022-05-13 09:19:28 PDT]    78B STANDARD XoBuIRDd2hOFVuvDieGO42nJTPz2mBQo v10 PUT foo.sh
[2022-05-13 09:19:27 PDT]    78B STANDARD bUtBobSO7Bo_FT4_aUC7G4EMu2hwDkds v9 PUT foo.sh
[2022-05-13 09:19:25 PDT]    78B STANDARD jSoB.Kp23bR1Xm9mPD9fkNFLTbDKHwOf v8 PUT foo.sh
[2022-05-13 09:19:22 PDT]    78B STANDARD SARv8bDhrCJQ1sJZX9gGDfXkAwlMaHJa v7 PUT foo.sh
[2022-05-13 09:19:02 PDT]    78B STANDARD FKBSFMaAUQ6vhMh368TQUS3gZ1.pm0mx v6 PUT foo.sh
[2022-05-13 09:19:00 PDT]    78B STANDARD IpNvlLtOReyempzCDo4c1wqLxMGsde87 v5 PUT foo.sh
[2022-05-13 09:17:15 PDT]    78B STANDARD uIwhknm3gGKZsMit.7WTgLHUU87IGiWE v4 PUT foo.sh
[2022-05-13 09:17:14 PDT]    78B STANDARD DRi9_isJnrrvvTeRqCVdtYevHPl3E5Nx v3 PUT foo.sh
[2022-05-13 09:17:13 PDT]    78B STANDARD S8ZlZPoWDw8_2lVMS9_1242VpAOK.jTq v2 PUT foo.sh
[2022-05-13 09:17:10 PDT]    78B STANDARD vDPN4EFugmdmV5jIuVSmDs.M9LTVLIYL v1 PUT foo.sh

❯ mc ls --versions s3/kpine-feebf0c1 | awk '{print $6}'
XoBuIRDd2hOFVuvDieGO42nJTPz2mBQo
bUtBobSO7Bo_FT4_aUC7G4EMu2hwDkds
jSoB.Kp23bR1Xm9mPD9fkNFLTbDKHwOf
SARv8bDhrCJQ1sJZX9gGDfXkAwlMaHJa
FKBSFMaAUQ6vhMh368TQUS3gZ1.pm0mx
IpNvlLtOReyempzCDo4c1wqLxMGsde87
uIwhknm3gGKZsMit.7WTgLHUU87IGiWE
DRi9_isJnrrvvTeRqCVdtYevHPl3E5Nx
S8ZlZPoWDw8_2lVMS9_1242VpAOK.jTq
vDPN4EFugmdmV5jIuVSmDs.M9LTVLIYL

❯ mc ls --versions s3/kpine-feebf0c1 | awk '{print $6}' | sort -c
sort: -:4: disorder: SARv8bDhrCJQ1sJZX9gGDfXkAwlMaHJa

❯ mc ls --versions s3/kpine-feebf0c1 | awk '{print $6}' | sort -c -r
sort: -:2: disorder: bUtBobSO7Bo_FT4_aUC7G4EMu2hwDkds
```

As you can see, there's no ordering to the version IDs. From our AWS-external viewpoint, the version IDs are random. You can observe this on your own with both AWS and Min.io. The latter uses UUID v4 version IDs, which are completely random.

Reply:
@keithpine, There is no requirement for having lexicographical sorted version IDs, the requirement is to have all the versions to be present in sorted order based on the creation time from latest to oldest, in the list-object-versions's output. And in order to achieve this we are making use of the motr kv store's capability to return the list of keys in  lexicographical sorted order, which will in turn make the list sorted based on timestamp because the version-ids are based on timestamp.